### PR TITLE
[2201.2.x] Implement change sub type of raw template code action

### LINKED ISSUE #34409: Add code action for convert to the subtypes of the `RawTemplate`
**Description:**
Consider the following source snippet. Here we can have a code action to convert the template literal to a string.

https://user-images.githubusercontent.com/1329674/146734608-54f4e3be-c652-41c8-9e41-12f496fe7525.mp4

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
This CA will use to resolve diagnostic for the code `BCE3936`. 

Fixes #34409

Master PR #38242

## Samples
https://user-images.githubusercontent.com/46857198/187604408-22eaa49f-7c10-48a0-b407-122c2bdc5e32.mov

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by malinthar — CHANGES_REQUESTED** [review]
Shall we add test cases for the following cases as well.

1. We should add a negative test case for the following case
```Ballerina
configurable string userName = ?;

const string myConst = `sample template`;
```
case 2
```Ballerina
function testPositionalArg() {
    myFunction(`sample teamplate`);
}


function myFunction(string expr) {

}
```

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ChangeSubTypeOfRawTemplateCodeAction.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.Optional;
+import java.util.Set;
+
+/***
+ * 
+ * Code Action for adding subtype for a raw template.
+ * @since 2201.2.0
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public class ChangeSubTypeOfRawTemplateCodeAction implements DiagnosticBasedCodeActionProvider {
```
[the remark this answers is not in the document]
Comment (LakshanWeerasinghe) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ChangeSubTypeOfRawTemplateCodeAction.java:53]: `ConvertRawTemplateToXmlOrString` is better. I'll do.

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/ChangeSubTypeOfRawTemplateCodeActionTest.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    @Test(dataProvider = "codeaction-data-provider")
+    public void test(String config) throws IOException, WorkspaceDocumentException {
+        super.test(config);
+    }
+
+    @DataProvider(name = "codeaction-data-provider")
+    @Override
+    public Object[][] dataProvider() {
+        return new Object[][]{
+                {"config1.json"},
```
[the remark this answers is not in the document]
Comment (LakshanWeerasinghe) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/ChangeSubTypeOfRawTemplateCodeActionTest.java:42]: I have already added a description field to each case. Can you check again. @malinthar

**Review by gayalkuruppu — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/constants/CommandConstants.java [hunk 3]
```diff
@@ -169,4 +169,6 @@ public class CommandConstants {
     public static final String EXTRACT_TYPE = "Extract type";
 
     public static final String MAKE_ANNOT_DECL_CONST = "Convert '%s' to constant";
+
+    public static final String CHANGE_TO_SUBTYPE_OF_RAW_TEMPLATE_TITLE = "Change to '%s' sub type of row template";
```
[the remark this answers is not in the document]
Comment (gayalkuruppu) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/constants/CommandConstants.java:173]: Shall we also rename test case folder name from `raw-template` to something more matching with the code action title?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ChangeSubTypeOfRawTemplateCodeAction.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            } else if (typeSymbol.get().typeKind() == TypeDescKind.STRING) {
+                typeSet.add("string");
+            } else if (typeSymbol.get().typeKind() == TypeDescKind.XML) {
+                typeSet.add("xml");
+            } else {
+                typeSet.addAll(Set.of("string", "xml"));
+            }
+        } else {
+            typeSet.addAll(Set.of("string", "xml"));
+        }
```
[the remark this answers is not in the document]
Comment (gayalkuruppu) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ChangeSubTypeOfRawTemplateCodeAction.java:97]: I too have some doubts, let's take this scenario
```
function testFunctionWithNoReturn() {
    return `Ballerina`;
}
```
1. Here we provide a quick fix code action but it does not fix the given diagnostic.
2. Here we provide both `string` and `xml` options(when there is no ret type desc), but when the return type is let's say `int|string` we do not provide `xml`.

We should either be flexible and provide both suggestions and make it a code action with no code action kind. Or else, we should stick to fixing the diagnostic and restrict cases which does not. (IMO first approach is better and agree with @malinthar)

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ConvertToXmlOrStringTemplateCodeAction.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+/***
+ *
+ * Code Action for adding subtype for a raw template.
```
Comment (mindula) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ConvertToXmlOrStringTemplateCodeAction.java:47]: Shall we move this description to line number 46?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ConvertToXmlOrStringTemplateCodeAction.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        return DIAGNOSTIC_CODE.equals(diagnostic.diagnosticInfo().code()) &&
+                CodeActionNodeValidator.validate(context.nodeAtRange()) && context.currentSemanticModel().isPresent()
+                && positionDetails.matchedNode().kind() == SyntaxKind.RAW_TEMPLATE_EXPRESSION;
+    }
+
+    @Override
+    public List<CodeAction> getCodeActions(Diagnostic diagnostic,
+                                           DiagBasedPositionDetails positionDetails,
+                                           CodeActionContext context) {
+
```
Comment (mindula) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ConvertToXmlOrStringTemplateCodeAction.java:68]: Let's remove this additional line

**Review by gayalkuruppu — APPROVED** [review]
LGTM

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/ConvertToXmlOrStringTemplateCodeActionTest.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import org.ballerinalang.langserver.commons.workspace.WorkspaceDocumentException;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+
+/**
+ * Test cases for @link{ConvertToXmlOrStringTemplateCodeAction}.
+ *
+ * @since 2201.2.0
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/ConvertToXmlOrStringTemplateCodeActionTest.java:28]: The version should be changed to 2201.2.3

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ChangeSubTypeOfRawTemplateCodeAction.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ChangeSubTypeOfRawTemplateCodeAction.java]: +1

**Review by malinthar — CHANGES_REQUESTED** [review]
Let's fix the failing tests

**Comment by IMS94** [comment]
Shall we send this to 2201.3.x instead?

File: language-server/modules/langserver-core/src/test/resources/codeaction/convert-to-str-xml-tmplt/config/config1.json [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+              "line": 0,
+              "character": 51
+            }
+          },
+          "newText": "xml `Module Var Declaration`"
+        }
+      ]
+    }
+  ],
+  "description": "Test module variable declaration with a row template"
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/test/resources/codeaction/convert-to-str-xml-tmplt/config/config1.json:28]:
```suggestion
  "description": "Test module variable declaration with a raw template"
```

File: language-server/modules/langserver-core/src/test/resources/codeaction/convert-to-str-xml-tmplt/config/config5.json [hunk 9]
```diff
@@ -0,0 +1,29 @@
+{
+  "position": {
+    "line": 15,
+    "character": 22
+  },
+  "description": "Test variable declaration of a row template when expecting a string",
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/test/resources/codeaction/convert-to-str-xml-tmplt/config/config5.json:6]:
```suggestion
  "description": "Test variable declaration of a raw template when expecting a string",
```

File: language-server/modules/langserver-core/src/test/resources/codeaction/convert-to-str-xml-tmplt/config/config6.json [hunk 10]
```diff
@@ -0,0 +1,48 @@
+{
+  "position": {
+    "line": 19,
+    "character": 15
+  },
+  "description": "Test variable declaration of a row template when expecting a int",
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/test/resources/codeaction/convert-to-str-xml-tmplt/config/config6.json:6]:
```suggestion
  "description": "Test variable declaration of a raw template when expecting a int",
```

File: language-server/modules/langserver-core/src/test/resources/codeaction/convert-to-str-xml-tmplt/config/config8.json [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+              "line": 28,
+              "character": 40
+            }
+          },
+          "newText": "string `sample template`"
+        }
+      ]
+    }
+  ],
+  "description": "Test const declaration which expecting a string type"
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/test/resources/codeaction/convert-to-str-xml-tmplt/config/config8.json:28]:
```suggestion
  "description": "Test const declaration when expecting a string type"
```

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ConvertToXmlOrStringTemplateCodeAction.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    @Override
+    public List<CodeAction> getCodeActions(Diagnostic diagnostic,
+                                           DiagBasedPositionDetails positionDetails,
+                                           CodeActionContext context) {
+        TemplateExpressionNode templateExpressionNode = (TemplateExpressionNode) positionDetails.matchedNode();
+        Optional<TypeSymbol> typeSymbol = templateExpressionNode.apply(new CodeActionContextTypeResolver(context));
+
+        Set<String> typeSet = new HashSet<>();
+        if (typeSymbol.isPresent()) {
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ConvertToXmlOrStringTemplateCodeAction.java:71]: Shall we use an early return if typeSymbol is not present?

**Review by TharushiJay — CHANGES_REQUESTED** [review]
Can you check the following case where we're not getting the code action `Convert to 'xml' template` for parameter `i`?

```
function test(xml i = `abc`) returns string {
    return string `def`;
}
```

**Comment by malinthar** [comment]
> Shall we send this to 2201.3.x instead?

Agreed!

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ConvertToXmlOrStringTemplateCodeAction.java (hunk 12 above)
Comment (LakshanWeerasinghe) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ConvertToXmlOrStringTemplateCodeAction.java:71]: In here the typeSymbol present means we can get the expected type of the raw template. There are cases where we cannot determine the expected type of the raw template. In those case we provide two code actions. Hence we cannot return an empty list even if the typeSymbol is not present.

example:
```
function typeSymbolNotPresent() {
     return `Ballerina`;
}
```

Comment (LakshanWeerasinghe) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/changetype/ConvertToXmlOrStringTemplateCodeAction.java:71]: @TharushiJay refer #38242 PR for the changes.

**Comment by LakshanWeerasinghe** [comment]
> Can you check the following case where we're not getting the code action `Convert to 'xml' template` for parameter `i`?
> 
> ```
> function test(xml i = `abc`) returns string {
>     return string `def`;
> }
> ```

Addressed this scenario.

**Comment by IMS94** [comment]
Closing this since we won't be sending this to 2201.2.x

**Pull request closed without merging**